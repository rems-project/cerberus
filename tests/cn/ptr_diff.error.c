int live_owned_footprint(char *p, char *q)
/*@
 requires
    take P = Owned<int[11]>(array_shift<char>(p, -2i64));
    ptr_eq(q, array_shift<char>(p, 12i64));
ensures
    take P2 = Owned<int[11]>(array_shift<char>(p, -2i64));
    P == P2;
    return == 12i32;
@*/
{
  // will fail without -- /*@ extract Owned<int>, 7u64; @*/
  return q - p;
}

int main(void)
{
    int arr[11] = { 0 };
    char *p = (char*) arr;
    live_owned_footprint(p + 2, p + 14);
}
