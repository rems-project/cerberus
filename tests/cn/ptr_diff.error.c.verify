return code: 1
[1/2]: live_RW_footprint -- fail
[2/2]: main -- pass
tests/cn/ptr_diff.error.c:13:10: error: Pointer `q` needs to be live for pointer difference
  return q - p;
         ~~^~~ 
Need an Alloc or RW in context with same allocation id
State file: file:///tmp/state__ptr_diff.error.c__live_RW_footprint.html
