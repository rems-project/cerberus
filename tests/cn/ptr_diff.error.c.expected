return code: 1
[1/2]: live_owned_footprint -- fail
[2/2]: main -- pass
tests/cn/ptr_diff.error.c:13:10: error: Pointer `q` needs to be live for pointer difference
  return q - p;
         ~~^~~ 
Need an Alloc or Owned in context with same allocation id
State file: file:///tmp/state__ptr_diff.error.c__live_owned_footprint.html
